python/cudf/cudf/core/abc.py

```diff
@@ -97,7 +97,14 @@ def device_serialize(self):
         """
         header, frames = self.serialize()
         assert all(
-            isinstance(f, (cudf.core.buffer.DeviceBufferLike, memoryview))
+            isinstance(
+                f,
+                (
+                    cudf.core.buffer.Buffer,
```

Comment:
Performing an `isinstance(obj, DeviceBufferLike)` seems to be different from `isinstance(obj, Buffer)`. The former is actually performing a check for `hasattr(obj, '__cuda_array_interface__')` which is triggering a call to the property, thus triggering detach process un-intentionally. To avoid that I had to replace it with a `isinstance(obj, Buffer)` which doesn't seem to trigger the `hasattr` check.

Reply:
Ah ok, yes. Because for protocols `isinstance` is just checking for conformity to an interface so is not via inheritance I guess.

This is a very non-obvious problem though. So it's easy to imagine that someone will later introduce an `isinstance(foo, DeviceBufferLike)` or similar which will kick this off.